File: scala/src/main/scala/rays/existential/math/Ray.scala
package rays.existential.math

/** A half-line that extends indefinitely in one direction.
  *
  * @param origin The ray's origin.
  * @param direction The direction pointed by the ray.
  */
final case class Ray(origin: Vector3, direction: Vector3):

  /** Returns `self` with its origin translated by `delta`. */
  def translatedBy(delta: Vector3): Ray =
    Ray(origin + delta, direction)

  /** Returns `self` rotated by `alpha`. */
  def rotatedBy(alpha: Quaternion): Ray =
    Ray(origin, direction * alpha)

end Ray

File: swift/Rays/Inheritance/Math/Ray.swift
/// A half-line that extends indefinitely in one direction.
public struct Ray: Hashable {

  /// The ray's origin.
  public var origin: Vector3

  /// The ray's direction, as a unit vector.
  public var direction: Vector3

  /// Initializes a ray with its origin and the direction toward which it points.
  public init(origin: Vector3, direction: Vector3) {
    self.origin = origin
    self.direction = direction
  }

  /// Returns `self` with its origin translated by `delta`.
  public func translated(by delta: Vector3) -> Ray {
    .init(origin: origin + delta, direction: direction)
  }

  /// Returns `self` rotated by `alpha`.
  public func rotated(by alpha: Quaternion) -> Ray {
    .init(origin: origin, direction: direction * alpha)
  }

}
